cmd/thanos/query.go

```diff
@@ -210,21 +225,84 @@ func runQuery(
 	})
 	reg.MustRegister(duplicatedStores)
 
-	dialOpts, err := extgrpc.StoreClientGRPCOpts(logger, reg, tracer, secure, cert, key, caCert, serverName)
-	if err != nil {
-		return errors.Wrap(err, "building gRPC client")
+	var storesConfig []store.Config
+	if len(storeConfigYAML) > 0 {
+		var err error
+		storesConfig, err = store.LoadConfig(storeConfigYAML)
+		if err != nil {
+			return errors.Wrap(err, "loading store config")
+		}
+	} else {
+		for _, addr := range storeAddrs {
+			if addr == "" {
+				return errors.New("static store address cannot be empty")
+			}
+		}
+
+		endpointsConfig := store.EndpointsConfig{
+			StaticAddresses: storeAddrs,
+		}
+		if fileSDConfig != nil {
+			endpointsConfig.FileSDConfigs = []file.SDConfig{*fileSDConfig}
+		}
+		storeConfig := store.Config{
+			Name:            "default",
+			EndpointsConfig: endpointsConfig,
+		}
+		if secure {
+			storeConfig.TlsConfig = &store.TlsConfig{
+				Cert:       cert,
+				Key:        key,
+				CaCert:     caCert,
+				ServerName: serverName,
+			}
+		}
+
+		storesConfig = []store.Config{storeConfig}
 	}
 
-	fileSDCache := cache.New()
-	dnsProvider := dns.NewProvider(
-		logger,
-		extprom.WrapRegistererWithPrefix("thanos_querier_store_apis_", reg),
-		dns.ResolverType(dnsSDResolver),
+	var storeSets []*query.StoreSet
+	var (
+		allClients = func() []store.Client {
+			var ret []store.Client
+			for _, ss := range storeSets {
+				ret = append(ret, ss.Get()...)
+			}
+			return ret
+		}
+		proxy            = store.NewProxyStore(logger, reg, allClients, component.Query, selectorLset, storeResponseTimeout)
+		queryableCreator = query.NewQueryableCreator(logger, proxy)
+		engine           = promql.NewEngine(
+			promql.EngineOpts{
+				Logger:        logger,
+				Reg:           reg,
+				MaxConcurrent: maxConcurrentQueries,
+				// TODO(bwplotka): Expose this as a flag: https://github.com/thanos-io/thanos/issues/703.
+				MaxSamples: math.MaxInt32,
+				Timeout:    queryTimeout,
+			},
+		)
 	)
```

Comment:
ok, played around locally, i suggest to move the whole `allClients` var block down after the for block in line 287, so we:
1. initialize `storesConfig`
2. initialize `storeSets`
3. initialize `proxy`, `queryableCreator`, and `engine`

that gives it (at least locally) a more sequential initialization flow and is easier to reason about, wdyt?

Reply:
yup, makes a lot of sense. although when I did it locally it made no sense having a var block then a bunch of variables initialised with the probers and then the code block below, so I combined the prober initialisation into the var block too. looks cleaner IMO.